src/main/java/bio/terra/workspace/app/configuration/spring/TraceInterceptorConfig.java

```diff
@@ -0,0 +1,113 @@
+  public static final String MDC_CORRELATION_ID_HEADER = "X-Correlation-ID";
+
+  private final Hashids hashids = new Hashids("requestIdSalt", 8);
+
+  @Autowired
+  public TraceInterceptorConfig(
+      TracingConfiguration tracingConfiguration, Environment environment) {
+    try {
+      StackdriverTraceExporter.createAndRegister(
+          StackdriverTraceConfiguration.builder()
+              .setProjectId(tracingConfiguration.getProjectId())
+              .setCredentials(
+                  ServiceAccountCredentials.fromStream(
+                      new FileInputStream(tracingConfiguration.getSaPath())))
+              .setFixedAttributes(
+                  Map.of(
+                      "/component",
+                      AttributeValue.stringAttributeValue(
+                          environment.getRequiredProperty("spring.application.name"))))
+              .build());
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+
+    TraceConfig globalTraceConfig = Tracing.getTraceConfig();
+    globalTraceConfig.updateActiveTraceParams(
+        globalTraceConfig.getActiveTraceParams().toBuilder()
+            .setSampler(Samplers.probabilitySampler(tracingConfiguration.getProbability()))
+            .build());
+  }
+
+  @Override
+  public void addInterceptors(InterceptorRegistry registry) {
+    registry.addInterceptor(
+        new HandlerInterceptor() {
+          @Override
+          public boolean preHandle(
+              HttpServletRequest httpRequest, HttpServletResponse httpResponse, Object handler)
+              throws Exception {
+
+            // get an mdc id from the request (if not found, create one), and pass it along in the
+            // response
+            String requestId = getMDCRequestId(httpRequest);
+            MDC.put(MDC_REQUEST_ID_KEY, requestId);
+            httpResponse.addHeader(MDC_REQUEST_ID_HEADER, requestId);
+
+            // add tags to Stackdriver traces
+            Tracing.getTracer()
+                .getCurrentSpan()
+                .putAttributes(
+                    Map.of(
+                        MDC_REQUEST_ID_KEY,
+                        AttributeValue.stringAttributeValue(requestId),
+                        "route",
+                        AttributeValue.stringAttributeValue(
+                            ((HandlerMethod) handler)
+                                .getMethodAnnotation(RequestMapping.class)
+                                .path()[0])));
+
+            return true;
+          }
+        });
+  }
+
+  private String generateRequestId() {
+    long generatedLong = ThreadLocalRandom.current().nextLong(0, Integer.MAX_VALUE);
+
+    return hashids.encode(generatedLong);
+  }
+
+  private String getMDCRequestId(HttpServletRequest httpRequest) {
+    return firstNonNull(
+        httpRequest.getHeader(MDC_REQUEST_ID_HEADER),
+        httpRequest.getHeader(MDC_CORRELATION_ID_HEADER),
+        generateRequestId());
+  }
```

Comment:
I _believe_ this is going to execute the relatively expensive `generateRequestId()` every time, then pass the resultant value into `firstNonNull` and discard it if the request has one of the desired headers. Could you rewrite this for efficiency so that it only calls `generateRequestId()` when it really needs to?